pysrc/utils/sampling_methods.py

```diff
@@ -13,72 +13,106 @@ def random_sample(data: list[list[str]],
     tmp = pd.Series(data)
     tmp = replaceEmptyField(tmp)
     tmp.dropna(inplace=True)
-    return tmp.sample(n=num_samples)
+    tmp = tmp.loc[tmp.astype(str).drop_duplicates().index]
 
+    if len(tmp)-1 > num_samples:
```

Comment:
I think you can just drop the `-1` as having `len(tmp) == num_samples` would mean there is no need to sample anything (the whole dataset will be returned either way):

```suggestion
    if len(tmp) > num_samples:
```

In case this is not the intended behavior, I'd trade the `-1` for `>=`:

```suggestion
    if len(tmp) >= num_samples:
```

I think this would be easier to understand.

This also applies to all the other methods below.